File: solutions/haskell/sum-of-multiples/3/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.Foldable (foldl')
import Data.Set (Set)
import qualified Data.Set as Set

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum multiples
  where
    multiples = foldl' f Set.empty factors
    f set factor
      | factor <= 0 = set
      | otherwise = Set.union set (Set.fromList . lst $ factor)
    lst factor = [i * factor | i <- takeWhile (\i -> i * factor < limit) [1..]]

File: solutions/kotlin/sum-of-multiples/2/src/main/kotlin/SumOfMultiples.kt
import kotlin.math.abs

object SumOfMultiples {
    fun sum(factors: Set<Int>, limit: Int): Int = factors
        .filter { it != 0 }
        .map { abs(it) }
        .flatMap { factor ->
            (1..((limit - 1) / factor))
                .map { factor * it }
        }
        .distinct()
        .sum()
}